File: smart-agriculture-suite/src/AnomalyModule.cpp
#include "AnomalyModule.h"

#include <math.h>
#include <string.h>

#include "AppConfig.h"

namespace agri {

namespace {
constexpr float kZScoreThreshold = 2.5f;
constexpr int kStuckThreshold = 30;
constexpr int kWindowSize = 60;
constexpr int kMaxNodesPerTree = 255;
}  // namespace

const char* AnomalyModule::alertLevelName(AlertLevel level) {
    switch (level) {
        case AlertLevel::Info: return "info";
        case AlertLevel::Warning: return "warning";
        case AlertLevel::Critical: return "critical";
        case AlertLevel::None:
        default:
            return "normal";
    }
}

void AnomalyModule::begin(int buzzerPin) {
    buzzerPin_ = buzzerPin;
    if (buzzerPin_ >= 0) {
        pinMode(buzzerPin_, OUTPUT);
        digitalWrite(buzzerPin_, LOW);
    }

    const char* names[kFeatureCount] = {"AirTemp", "AirHumi", "SoilHumi", "Light"};
    const char* labels[kFeatureCount] = {"AirTemp", "AirHumi", "Soil", "Light"};

    for (int i = 0; i < kFeatureCount; ++i) {
        sensors_[i].name = names[i];
        sensors_[i].label = labels[i];
    }
}

void AnomalyModule::update(const SensorSnapshot& snapshot, bool sampleUpdated, unsigned long nowMs) {
    if (sampleUpdated) {
        float values[kFeatureCount] = {
            snapshot.airTemp,
            snapshot.airHumi,
            snapshot.soilHumi,
            snapshot.lightValue
        };

        for (int i = 0; i < kFeatureCount; ++i) {
            checkSensorFault(sensors_[i], values[i]);
            updateSlidingWindow(sensors_[i], values[i]);
            sensors_[i].currentValue = values[i];
            calculateStats(sensors_[i]);

            if (values[i] < sensors_[i].minEver) {
                sensors_[i].minEver = values[i];
            }
            if (values[i] > sensors_[i].maxEver) {
                sensors_[i].maxEver = values[i];
            }
        }

        if (trainBufferCount_ < kTrainBufferSize) {
            for (int i = 0; i < kFeatureCount; ++i) {
                trainBuffer_[trainBufferCount_][i] = values[i];
            }
            ++trainBufferCount_;
        }

        detectZScoreAnomaly();
        updateAlertLevel();
        ++totalSamples_;
    }

    if (nowMs - lastIforestRunMs_ >= kAnomalyIsolationIntervalMs) {
        if (trainBufferCount_ >= 50 && !iforestTrained_) {
            buildIsolationForest();
        }
        if (iforestTrained_) {
            float features[kFeatureCount] = {
                sensors_[0].currentValue,
                sensors_[1].currentValue,
                sensors_[2].currentValue,
                sensors_[3].currentValue
            };
            iforestScore_ = runIsolationForest(features);
            if (iforestScore_ > 0.65f) {
                addAlert(AlertLevel::Warning, "IForest", "multi-sensor anomaly detected");
                ++totalAnomalies_;
            }
            updateAlertLevel();
        }
        lastIforestRunMs_ = nowMs;
    }
}

void AnomalyModule::clear() {
    alertCount_ = 0;
    alertIndex_ = 0;
    currentLevel_ = AlertLevel::None;
    totalAnomalies_ = 0;
    for (int i = 0; i < kFeatureCount; ++i) {
        sensors_[i].anomalyCount = 0;
        sensors_[i].isAnomalous = false;
        sensors_[i].isDisconnected = false;
        sensors_[i].isStuck = false;
    }
}

void AnomalyModule::updateSlidingWindow(SensorStats& stats, float value) {
    stats.window[stats.windowIndex] = value;
    stats.windowIndex = (stats.windowIndex + 1) % kWindowSize;
    if (stats.windowCount < kWindowSize) {
        ++stats.windowCount;
    }
}

void AnomalyModule::calculateStats(SensorStats& stats) {
    if (stats.windowCount < 3) {
        return;
    }

    float sum = 0.0f;
    for (int i = 0; i < stats.windowCount; ++i) {
        sum += stats.window[i];
    }
    stats.mean = sum / static_cast<float>(stats.windowCount);

    float sumSq = 0.0f;
    for (int i = 0; i < stats.windowCount; ++i) {
        const float diff = stats.window[i] - stats.mean;
        sumSq += diff * diff;
    }
    stats.stddev = sqrtf(sumSq / static_cast<float>(stats.windowCount));

    if (stats.stddev > 0.001f) {
        stats.lastZScore = (stats.currentValue - stats.mean) / stats.stddev;
    } else {
        stats.lastZScore = 0.0f;
    }
}

void AnomalyModule::checkSensorFault(SensorStats& stats, float value) {
    const bool isLight = strcmp(stats.name, "Light") == 0;

    if (!isLight && value <= 0.001f && stats.windowCount > 10) {
        if (!stats.isDisconnected) {
            addAlert(AlertLevel::Critical, stats.name, String(stats.label) + " disconnected");
        }
        stats.isDisconnected = true;
        return;
    }
    stats.isDisconnected = false;

    if (fabsf(value - stats.lastValue) < 0.01f) {
        ++stats.stuckCount;
        if (stats.stuckCount >= kStuckThreshold && !stats.isStuck) {
            stats.isStuck = true;
            addAlert(AlertLevel::Info, stats.name, String(stats.label) + " may be stuck");
        }
    } else {
        stats.stuckCount = 0;
        stats.isStuck = false;
    }
    stats.lastValue = value;
}

void AnomalyModule::detectZScoreAnomaly() {
    for (int i = 0; i < kFeatureCount; ++i) {
        if (sensors_[i].windowCount < 10) {
            continue;
        }

        const float absZ = fabsf(sensors_[i].lastZScore);
        if (absZ > kZScoreThreshold) {
            if (!sensors_[i].isAnomalous) {
                sensors_[i].isAnomalous = true;
                ++sensors_[i].anomalyCount;
                ++totalAnomalies_;

                String message = String(sensors_[i].label) + " z-score=" + String(sensors_[i].lastZScore, 2);
                addAlert(absZ > 3.5f ? AlertLevel::Critical : AlertLevel::Warning, sensors_[i].name, message);
            }
        } else {
            sensors_[i].isAnomalous = false;
        }
    }
}

void AnomalyModule::buildIsolationForest() {
    const int n = trainBufferCount_;
    if (n <= 1) {
        return;
    }

    for (int t = 0; t < kTrees; ++t) {
        forest_[t].nodeCount = 0;

        float minVals[kFeatureCount];
        float maxVals[kFeatureCount];
        for (int f = 0; f < kFeatureCount; ++f) {
            minVals[f] = 9999.0f;
            maxVals[f] = -9999.0f;
        }

        for (int i = 0; i < n; ++i) {
            for (int f = 0; f < kFeatureCount; ++f) {
                if (trainBuffer_[i][f] < minVals[f]) {
                    minVals[f] = trainBuffer_[i][f];
                }
                if (trainBuffer_[i][f] > maxVals[f]) {
                    maxVals[f] = trainBuffer_[i][f];
                }
            }
        }

        // Build tree recursively to full depth kDepth
        buildIsolationNode(t, minVals, maxVals, 0);
    }

    iforestTrained_ = true;
}

int AnomalyModule::buildIsolationNode(int treeIdx, float minVals[], float maxVals[], int depth) {
    if (depth >= kDepth || forest_[treeIdx].nodeCount >= kMaxNodesPerTree) {
        return -1;
    }

    const int nodeIndex = forest_[treeIdx].nodeCount++;

    int feature = random(kFeatureCount);
    float splitValue = minVals[feature] + random(1000) / 1000.0f * (maxVals[feature] - minVals[feature]);
    forest_[treeIdx].nodes[nodeIndex].splitFeature = feature;
    forest_[treeIdx].nodes[nodeIndex].splitValue = splitValue;
    forest_[treeIdx].nodes[nodeIndex].left = -1;
    forest_[treeIdx].nodes[nodeIndex].right = -1;

    float leftMin[kFeatureCount], leftMax[kFeatureCount];
    float rightMin[kFeatureCount], rightMax[kFeatureCount];
    memcpy(leftMin, minVals, sizeof(leftMin));
    memcpy(leftMax, maxVals, sizeof(leftMax));
    memcpy(rightMin, minVals, sizeof(rightMin));
    memcpy(rightMax, maxVals, sizeof(rightMax));

    // Left child: samples where value < splitValue, so max of split feature = splitValue
    leftMax[feature] = splitValue;
    // Right child: samples where value >= splitValue, so min of split feature = splitValue
    rightMin[feature] = splitValue;

    int leftIdx = buildIsolationNode(treeIdx, leftMin, leftMax, depth + 1);
    int rightIdx = buildIsolationNode(treeIdx, rightMin, rightMax, depth + 1);

    forest_[treeIdx].nodes[nodeIndex].left = leftIdx;
    forest_[treeIdx].nodes[nodeIndex].right = rightIdx;

    return nodeIndex;
}

float AnomalyModule::runIsolationForest(float features[]) const {
    if (!iforestTrained_ || trainBufferCount_ <= 1) {
        return 0.0f;
    }

    float avgPath = 0.0f;
    for (int t = 0; t < kTrees; ++t) {
        avgPath += pathLength(t, 0, features, 0);
    }
    avgPath /= static_cast<float>(kTrees);

    const float c = averagePathLength(trainBufferCount_);
    if (c <= 0.0f) {
        return 0.0f;
    }
    return powf(2.0f, -(avgPath / c));
}

float AnomalyModule::pathLength(int treeIdx, int nodeIdx, float features[], int depth) const {
    if (depth >= kDepth || nodeIdx < 0 || nodeIdx >= forest_[treeIdx].nodeCount) {
        return static_cast<float>(depth);
    }

    const IForestNode& node = forest_[treeIdx].nodes[nodeIdx];
    if (node.left == -1 && node.right == -1) {
        return static_cast<float>(depth);
    }

    if (features[node.splitFeature] < node.splitValue) {
        return pathLength(treeIdx, node.left, features, depth + 1);
    }
    return pathLength(treeIdx, node.right, features, depth + 1);
}

float AnomalyModule::averagePathLength(int n) const {
    if (n <= 1) {
        return 1.0f;
    }
    const float h = logf(static_cast<float>(n - 1)) + 0.5772156649f;
    return 2.0f * h - 2.0f * (static_cast<float>(n - 1) / static_cast<float>(n));
}

void AnomalyModule::addAlert(AlertLevel level, const char* sensor, const String& message) {
    for (int i = 0; i < min(alertCount_, kMaxAlerts); ++i) {
        const int idx = (alertIndex_ - 1 - i + kMaxAlerts) % kMaxAlerts;
        if (millis() - alerts_[idx].timestamp < 5000UL && strcmp(alerts_[idx].sensor, sensor) == 0) {
            return;
        }
    }

    const int idx = alertIndex_ % kMaxAlerts;
    alerts_[idx].timestamp = millis();
    alerts_[idx].level = level;
    strncpy(alerts_[idx].sensor, sensor, sizeof(alerts_[idx].sensor) - 1);
    alerts_[idx].sensor[sizeof(alerts_[idx].sensor) - 1] = '\0';
    strncpy(alerts_[idx].message, message.c_str(), sizeof(alerts_[idx].message) - 1);
    alerts_[idx].message[sizeof(alerts_[idx].message) - 1] = '\0';

    alertIndex_ = (alertIndex_ + 1) % kMaxAlerts;
    ++alertCount_;

    beep(level);
}

void AnomalyModule::updateAlertLevel() {
    AlertLevel maxLevel = AlertLevel::None;
    for (int i = 0; i < kFeatureCount; ++i) {
        if (sensors_[i].isDisconnected) {
            maxLevel = AlertLevel::Critical;
            break;
        }
        if (sensors_[i].isAnomalous && maxLevel < AlertLevel::Warning) {
            maxLevel = AlertLevel::Warning;
        }
        if (sensors_[i].isStuck && maxLevel < AlertLevel::Info) {
            maxLevel = AlertLevel::Info;
        }
    }
    if (iforestScore_ > 0.65f && maxLevel < AlertLevel::Warning) {
        maxLevel = AlertLevel::Warning;
    }
    currentLevel_ = maxLevel;
}

void AnomalyModule::beep(AlertLevel level) const {
    if (buzzerPin_ < 0 || level == AlertLevel::None) {
        return;
    }

    int count = 1;
    int onMs = 60;
    int offMs = 60;

    if (level == AlertLevel::Warning) {
        count = 2;
        onMs = 100;
    } else if (level == AlertLevel::Critical) {
        count = 4;
        onMs = 50;
        offMs = 50;
    }

    for (int i = 0; i < count; ++i) {
        digitalWrite(buzzerPin_, HIGH);
        delay(onMs);
        digitalWrite(buzzerPin_, LOW);
        delay(offMs);
    }
}

void AnomalyModule::writeStatus(JsonDocument& doc) const {
    doc["alertLevel"] = static_cast<int>(currentLevel_);
    doc["alertLevelName"] = alertLevelName(currentLevel_);
    doc["totalSamples"] = totalSamples_;
    doc["totalAnomalies"] = totalAnomalies_;
    doc["iforestTrained"] = iforestTrained_;
    doc["iforestScore"] = iforestScore_;

    JsonArray sensors = doc["sensors"].to<JsonArray>();
    for (int i = 0; i < kFeatureCount; ++i) {
        JsonObject sensor = sensors.add<JsonObject>();
        sensor["name"] = sensors_[i].name;
        sensor["label"] = sensors_[i].label;
        sensor["value"] = sensors_[i].currentValue;
        sensor["mean"] = sensors_[i].mean;
        sensor["stddev"] = sensors_[i].stddev;
        sensor["zScore"] = sensors_[i].lastZScore;
        sensor["isAnomalous"] = sensors_[i].isAnomalous;
        sensor["isStuck"] = sensors_[i].isStuck;
        sensor["isDisconnected"] = sensors_[i].isDisconnected;
        sensor["anomalyCount"] = sensors_[i].anomalyCount;
        sensor["min"] = sensors_[i].minEver;
        sensor["max"] = sensors_[i].maxEver;
    }
}

void AnomalyModule::writeAlerts(JsonDocument& doc) const {
    JsonArray alerts = doc.to<JsonArray>();
    const int count = min(alertCount_, kMaxAlerts);
    for (int i = 0; i < count; ++i) {
        const int idx = (alertIndex_ - 1 - i + kMaxAlerts) % kMaxAlerts;
        JsonObject alert = alerts.add<JsonObject>();
        alert["timestamp"] = alerts_[idx].timestamp;
        alert["level"] = static_cast<int>(alerts_[idx].level);
        alert["levelName"] = alertLevelName(alerts_[idx].level);
        alert["sensor"] = alerts_[idx].sensor;
        alert["message"] = alerts_[idx].message;
    }
}

bool AnomalyModule::writeSensorDetail(const String& sensorName, JsonDocument& doc) const {
    for (int i = 0; i < kFeatureCount; ++i) {
        if (sensorName != sensors_[i].name) {
            continue;
        }

        doc["name"] = sensors_[i].name;
        doc["label"] = sensors_[i].label;
        doc["value"] = sensors_[i].currentValue;
        doc["mean"] = sensors_[i].mean;
        doc["stddev"] = sensors_[i].stddev;
        doc["zScore"] = sensors_[i].lastZScore;

        JsonArray window = doc["window"].to<JsonArray>();
        for (int w = 0; w < sensors_[i].windowCount; ++w) {
            const int index = (sensors_[i].windowIndex - sensors_[i].windowCount + w + kWindowSize) % kWindowSize;
            window.add(sensors_[i].window[index]);
        }
        return true;
    }
    return false;
}

}  // namespace agri

File: smart-agriculture-atlas200dk/src/ai/anomaly_module.py
"""
智润智慧农业套件 - Atlas 200I DK A2 版
异常检测模块 - 三层异常检测: 滑动窗口 + Z-Score + Isolation Forest

对应原ESP32项目的 AnomalyModule.h/AnomalyModule.cpp
Python版使用NumPy加速, 算法逻辑完全一致
"""

import math
import time
import random
import logging
from dataclasses import dataclass, field
from typing import List, Optional
from collections import deque

import numpy as np

from config.app_types import SensorSnapshot, AnomalyLevel, AnomalyResult

logger = logging.getLogger(__name__)


@dataclass
class Alert:
    timestamp: float = 0.0
    level: AnomalyLevel = AnomalyLevel.Normal
    sensor: str = ""
    message: str = ""


@dataclass
class SensorStats:
    name: str = ""
    label: str = ""
    current_value: float = 0.0
    last_value: float = 0.0
    mean: float = 0.0
    stddev: float = 0.0
    last_z_score: float = 0.0
    is_anomalous: bool = False
    is_stuck: bool = False
    is_disconnected: bool = False
    stuck_count: int = 0
    anomaly_count: int = 0
    min_ever: float = float('inf')
    max_ever: float = float('-inf')
    window: deque = field(default_factory=lambda: deque(maxlen=60))


@dataclass
class IForestNode:
    split_feature: int = 0
    split_value: float = 0.0
    left: int = -1
    right: int = -1


@dataclass
class IForestTree:
    nodes: List[IForestNode] = field(default_factory=list)


class AnomalyModule:
    """三层异常检测: 滑动窗口统计 + Z-Score + Isolation Forest"""

    FEATURE_COUNT = 4
    WINDOW_SIZE = 60
    STUCK_THRESHOLD = 30
    ZSCORE_WARN = 2.5
    ZSCORE_CRITICAL = 3.5
    IFOREST_THRESHOLD = 0.65
    MAX_TREES = 10
    MAX_DEPTH = 8
    MAX_NODES_PER_TREE = 255
    TRAIN_BUFFER_SIZE = 200
    MIN_BUILD_SAMPLES = 50
    MAX_ALERTS = 20

    def __init__(self, buzzer=None):
        self._buzzer = buzzer  # ActuatorController reference

        self._sensor_names = ["AirTemp", "AirHumi", "SoilHumi", "Light"]
        self._sensor_labels = ["AirTemp", "AirHumi", "Soil", "Light"]
        self._sensors = [
            SensorStats(name=n, label=l)
            for n, l in zip(self._sensor_names, self._sensor_labels)
        ]

        self._current_level = AnomalyLevel.Normal
        self._total_samples = 0
        self._total_anomalies = 0
        self._iforest_score = 0.0
        self._iforest_trained = False
        self._last_iforest_time = 0.0

        self._train_buffer = []
        self._train_buffer_count = 0

        self._forest: List[IForestTree] = []
        self._alerts: deque = deque(maxlen=self.MAX_ALERTS)

    def update(self, snapshot: SensorSnapshot, sample_updated: bool, now: float = 0.0):
        if now == 0.0:
            now = time.time()

        if sample_updated:
            values = [
                snapshot.air_temp,
                snapshot.air_humi,
                snapshot.soil_humi,
                snapshot.light_intensity,
            ]

            for i, val in enumerate(values):
                self._check_sensor_fault(self._sensors[i], val)
                self._sensors[i].window.append(val)
                self._sensors[i].current_value = val
                self._calculate_stats(self._sensors[i])

                if val < self._sensors[i].min_ever:
                    self._sensors[i].min_ever = val
                if val > self._sensors[i].max_ever:
                    self._sensors[i].max_ever = val

            if self._train_buffer_count < self.TRAIN_BUFFER_SIZE:
                self._train_buffer.append(values[:])
                self._train_buffer_count += 1

            self._detect_z_score_anomaly()
            self._update_alert_level()
            self._total_samples += 1

        # Isolation Forest每60秒运行一次
        if now - self._last_iforest_time >= 60.0:
            if self._train_buffer_count >= self.MIN_BUILD_SAMPLES and not self._iforest_trained:
                self._build_isolation_forest()
            if self._iforest_trained:
                features = [s.current_value for s in self._sensors]
                self._iforest_score = self._run_isolation_forest(features)
                if self._iforest_score > self.IFOREST_THRESHOLD:
                    self._add_alert(AnomalyLevel.Warning, "IForest",
                                    "multi-sensor anomaly detected")
                    self._total_anomalies += 1
                self._update_alert_level()
            self._last_iforest_time = now

    @property
    def current_level(self) -> AnomalyLevel:
        return self._current_level

    @property
    def iforest_score(self) -> float:
        return self._iforest_score

    @property
    def iforest_trained(self) -> bool:
        return self._iforest_trained

    @property
    def total_anomalies(self) -> int:
        return self._total_anomalies

    @property
    def total_samples(self) -> int:
        return self._total_samples

    def clear(self):
        self._alerts.clear()
        self._current_level = AnomalyLevel.Normal
        self._total_anomalies = 0
        for s in self._sensors:
            s.anomaly_count = 0
            s.is_anomalous = False
            s.is_disconnected = False
            s.is_stuck = False

    # ------------------------------------------------------------------
    # 滑动窗口统计
    # ------------------------------------------------------------------
    def _calculate_stats(self, stats: SensorStats):
        if len(stats.window) < 3:
            return
        arr = np.array(stats.window)
        stats.mean = float(np.mean(arr))
        stats.stddev = float(np.std(arr))
        if stats.stddev > 0.001:
            stats.last_z_score = (stats.current_value - stats.mean) / stats.stddev
        else:
            stats.last_z_score = 0.0

    def _check_sensor_fault(self, stats: SensorStats, value: float):
        is_light = stats.name == "Light"

        if not is_light and value <= 0.001 and len(stats.window) > 10:
            if not stats.is_disconnected:
                self._add_alert(AnomalyLevel.Critical, stats.name,
                                f"{stats.label} disconnected")
            stats.is_disconnected = True
            return
        stats.is_disconnected = False

        if abs(value - stats.last_value) < 0.01:
            stats.stuck_count += 1
            if stats.stuck_count >= self.STUCK_THRESHOLD and not stats.is_stuck:
                stats.is_stuck = True
                self._add_alert(AnomalyLevel.Normal, stats.name,
                                f"{stats.label} may be stuck")
        else:
            stats.stuck_count = 0
            stats.is_stuck = False
        stats.last_value = value

    def _detect_z_score_anomaly(self):
        for s in self._sensors:
            if len(s.window) < 10:
                continue
            abs_z = abs(s.last_z_score)
            if abs_z > self.ZSCORE_WARN:
                if not s.is_anomalous:
                    s.is_anomalous = True
                    s.anomaly_count += 1
                    self._total_anomalies += 1
                    level = AnomalyLevel.Critical if abs_z > self.ZSCORE_CRITICAL else AnomalyLevel.Warning
                    self._add_alert(level, s.name,
                                    f"{s.label} z-score={s.last_z_score:.2f}")
            else:
                s.is_anomalous = False

    # ------------------------------------------------------------------
    # Isolation Forest
    # ------------------------------------------------------------------
    def _build_isolation_forest(self):
        data = np.array(self._train_buffer[:self._train_buffer_count])
        n = data.shape[0]
        if n <= 1:
            return

        min_vals = data.min(axis=0)
        max_vals = data.max(axis=0)

        self._forest = []
        for _ in range(self.MAX_TREES):
            tree = IForestTree()
            self._build_isolation_node(tree, min_vals.copy(), max_vals.copy(), 0)
            self._forest.append(tree)

        self._iforest_trained = True
        logger.info(f"Isolation Forest构建完成: {self.MAX_TREES}棵树, {n}个训练样本")

    def _build_isolation_node(self, tree: IForestTree, min_vals: np.ndarray,
                               max_vals: np.ndarray, depth: int) -> int:
        if depth >= self.MAX_DEPTH or len(tree.nodes) >= self.MAX_NODES_PER_TREE:
            return -1

        node_index = len(tree.nodes)
        feature = random.randint(0, self.FEATURE_COUNT - 1)
        split_value = min_vals[feature] + random.random() * (max_vals[feature] - min_vals[feature])

        node = IForestNode(split_feature=feature, split_value=split_value)
        tree.nodes.append(node)

        left_max = max_vals.copy()
        left_max[feature] = split_value
        right_min = min_vals.copy()
        right_min[feature] = split_value

        left_idx = self._build_isolation_node(tree, min_vals.copy(), left_max, depth + 1)
        right_idx = self._build_isolation_node(tree, right_min.copy(), max_vals.copy(), depth + 1)

        tree.nodes[node_index].left = left_idx
        tree.nodes[node_index].right = right_idx
        return node_index

    def _run_isolation_forest(self, features: list) -> float:
        if not self._iforest_trained or self._train_buffer_count <= 1:
            return 0.0

        avg_path = 0.0
        for tree in self._forest:
            avg_path += self._path_length(tree, 0, features, 0)
        avg_path /= self.MAX_TREES

        c = self._average_path_length(self._train_buffer_count)
        if c <= 0.0:
            return 0.0
        return 2.0 ** (-(avg_path / c))

    def _path_length(self, tree: IForestTree, node_idx: int,
                      features: list, depth: int) -> float:
        if depth >= self.MAX_DEPTH or node_idx < 0 or node_idx >= len(tree.nodes):
            return float(depth)

        node = tree.nodes[node_idx]
        if node.left == -1 and node.right == -1:
            return float(depth)

        if features[node.split_feature] < node.split_value:
            return self._path_length(tree, node.left, features, depth + 1)
        return self._path_length(tree, node.right, features, depth + 1)

    @staticmethod
    def _average_path_length(n: int) -> float:
        if n <= 1:
            return 1.0
        h = math.log(n - 1) + 0.5772156649
        return 2.0 * h - 2.0 * ((n - 1) / n)

    # ------------------------------------------------------------------
    # 告警管理
    # ------------------------------------------------------------------
    def _add_alert(self, level: AnomalyLevel, sensor: str, message: str):
        now = time.time()
        # 5秒内同传感器不重复告警
        for alert in self._alerts:
            if now - alert.timestamp < 5.0 and alert.sensor == sensor:
                return

        self._alerts.append(Alert(
            timestamp=now, level=level, sensor=sensor, message=message
        ))
        self._beep(level)

    def _update_alert_level(self):
        max_level = AnomalyLevel.Normal
        for s in self._sensors:
            if s.is_disconnected:
                max_level = AnomalyLevel.Critical
                break
            if s.is_anomalous and max_level.value < AnomalyLevel.Warning.value:
                max_level = AnomalyLevel.Warning
            if s.is_stuck and max_level.value < AnomalyLevel.Normal.value:
                max_level = AnomalyLevel.Normal
        if self._iforest_score > self.IFOREST_THRESHOLD and max_level.value < AnomalyLevel.Warning.value:
            max_level = AnomalyLevel.Warning
        self._current_level = max_level

    def _beep(self, level: AnomalyLevel):
        if not self._buzzer or level == AnomalyLevel.Normal:
            return
        if level == AnomalyLevel.Warning:
            self._buzzer.beep(count=2, on_ms=100, off_ms=60)
        elif level == AnomalyLevel.Critical:
            self._buzzer.beep(count=4, on_ms=50, off_ms=50)

    def to_dict(self) -> dict:
        return {
            "alertLevel": self._current_level.value,
            "alertLevelName": self._current_level.name,
            "totalSamples": self._total_samples,
            "totalAnomalies": self._total_anomalies,
            "iforestTrained": self._iforest_trained,
            "iforestScore": round(self._iforest_score, 4),
            "sensors": [
                {
                    "name": s.name,
                    "label": s.label,
                    "value": round(s.current_value, 2),
                    "mean": round(s.mean, 2),
                    "stddev": round(s.stddev, 4),
                    "zScore": round(s.last_z_score, 4),
                    "isAnomalous": s.is_anomalous,
                    "isStuck": s.is_stuck,
                    "isDisconnected": s.is_disconnected,
                    "anomalyCount": s.anomaly_count,
                    "min": round(s.min_ever, 2) if s.min_ever != float('inf') else 0,
                    "max": round(s.max_ever, 2) if s.max_ever != float('-inf') else 0,
                }
                for s in self._sensors
            ],
        }
